### DISCUSSION TITLE
eval( ) issue

### DISCUSSION THREAD

**Discussion by ivanjeremic** [thekevinscott/UpscalerJS#381] [Ideas]
I try to use UpscalerJS in services like Cloudflare Workers and other functions providers but non of them allow JavaScript `eval()` and it seems like Upscaler is using it, can this be changed so Upscaler works without eval?

**Comment by thekevinscott** [comment]
Do you have an example of the error you’re seeing? Or perhaps a repo demonstrating the issue?

I don’t believe there’s eval in the upscaler code base, but maybe a downstream dependency is using it?

**Reply by thekevinscott** [reply]
More generally though to your point, I agree that `eval()` should be avoided! Cloudflare Workers are a great target to use UpscalerJS in. Is a Cloudflare Worker environment Node or Browser-based?

**Reply by ivanjeremic** [reply]
> More generally though to your point, I agree that `eval()` should be avoided! Cloudflare Workers are a great target to use UpscalerJS in. Is a Cloudflare Worker environment Node or Browser-based?

Cloudflare Workers are a custom thing based on just v8 by Cloudflare so not node and not Browser. There are many useful places this would be perfect, but the most useful like workers or other serverless/edge function providers do not allow `eval()` also if someone would like to use UpscalerJS inside a Browser Extension this would also not be possible because `eval()` is not allowed in Browser Extensions too. I have no repo but just the use of Upscaler `new Upscaler();` gives this errors on all platforms where `eval` is not allowed so it must be an important dependency.

**Reply by thekevinscott** [reply]
An example of UpscalerJS working with Cloudflare Workers is here:

https://github.com/thekevinscott/UpscalerJS/tree/main/examples/cloudflare-worker

**Reply by ivanjeremic** [reply]
Awesome thanks!

**Comment by thekevinscott** [comment]
Do you have an example of the error you’re seeing? Or perhaps a repo demonstrating the issue?

I don’t believe there’s eval in the upscaler code base, but maybe a downstream dependency is using it?

**Reply by ivanjeremic** [reply]
Yes "seedrandom" is using it 

```
Uncaught EvalError: Code generation from strings disallowed for this context
  at line 2
  at line 2 in ./node_modules/seedrandom/seedrandom.js
  at line 2 in s
  at line 2 in ./node_modules/seedrandom/index.js
  at line 2 in s
  at line 2 in ./node_modules/@tensorflow/tfjs-data/dist/dataset.js
  at line 2 in s
  at line 2 in ./node_modules/@tensorflow/tfjs-data/dist/index.js
  at line 2 in s
  at line 2 in ./node_modules/@tensorflow/tfjs/dist/index.js
```

**Reply by thekevinscott** [reply]
I created https://github.com/thekevinscott/UpscalerJS/issues/443 to track work on supporting Cloudflare Workers.

Like you mentioned above, the Cloudflare environment is neither browser nor Node, but is closest to a subset of the browser environment.

It looks like you're trying to import `@tensorflow/tfjs-node`, which won't work; like you've discovered, `seedrandom` is the package that's got the `eval` call and that is required by `@tensorflow/tfjs-node`.

What you can do instead is use the browser version with `@tensorflow/tfjs`. However, a large limitation is that the environment lacks access to `HTMLImageElement` and `Canvas`, which are used to handle some of the base64 processing functions. 

What this means is that you'll have to provide an input as a tensor, and receive output as a tensor too; you won't be able to base64 encode it.

You can see an example (still being worked on) here: https://github.com/thekevinscott/UpscalerJS/tree/main/docs/workers/upscale

It's on my list (#443) to investigate this further and see if we can enable base64 encoding in a Cloudflare environment.
